locator/ec2_snitch.cc

```diff
@@ -63,17 +64,26 @@ future<> ec2_snitch::start() {
     });
 }
 
-future<sstring> ec2_snitch::aws_api_call(sstring addr, uint16_t port, sstring cmd) {
+future<sstring> ec2_snitch::aws_api_call(sstring addr, uint16_t port, sstring cmd, std::optional<sstring> token) {
     return connect(socket_address(inet_address{addr}, port))
-    .then([this, addr, cmd] (connected_socket fd) {
+    .then([this, addr, cmd, token] (connected_socket fd) {
         _sd = std::move(fd);
         _in = _sd.input();
         _out = _sd.output();
-        _zone_req = sstring("GET ") + cmd +
-                    sstring(" HTTP/1.1\r\nHost: ") +addr +
-                    sstring("\r\n\r\n");
 
-        return _out.write(_zone_req.c_str()).then([this] {
+        if (token) {
+            _req = sstring("GET ") + cmd +
+                   sstring(" HTTP/1.1\r\nHost: ") +addr +
+                   sstring("\r\nX-aws-ec2-metadata-token: ") + *token +
+                   sstring("\r\n\r\n");
+        } else {
+            _req = sstring("PUT ") + cmd +
+                   sstring(" HTTP/1.1\r\nHost: ") + addr +
+                   sstring("\r\nX-aws-ec2-metadata-token-ttl-seconds: 21600") +
```

Comment:
Why do you need the token to be alive for 6 hours? You don't cache it AFAIKT and will use it only in the context of the snitch initialization.

Doc pages you referenced above are reusing the pre-generated token where it makes sense for it to last that long. Here 6h seems to be excessive - a few 10s of seconds should be more than enough IMO.

Reply:
Actually most examples are 21600 which is the maximum allowed. It is not like you have a token limit however, so you can get as many tokens as you want. What AWS does LIMIT us is on the number of API concurrent requests, and this is regardless if you have IMDS or IMDSv2.

That said, I agree it is high and we don't need it, and since the primary reason for users to rely on IMDSv2 is to avoid SSRF calls it makes sense for us to reduce it so we can let the token expire sooner. :-)